validators/cgroup_validator.go

```diff
@@ -82,6 +103,21 @@ func (c *CgroupsValidator) validateCgroupSubsystems(cgroups, subsystems []string
 }
 
 func (c *CgroupsValidator) getCgroupSubsystems() ([]string, error) {
+	// /proc/cgroups is meaningless for v2
+	// https://github.com/torvalds/linux/blob/v5.3/Documentation/admin-guide/cgroup-v2.rst#deprecated-v1-core-features
+	if IsCgroup2UnifiedMode() {
+		// "pseudo" controllers do not appear in /sys/fs/cgroup/cgroup.controllers.
+		// - devices: implemented in kernel 4.15
+		// - freezer: implemented in kernel 5.2
+		// We assume these are always available, as it is hard to detect availability.
+		pseudo := []string{"devices", "freezer"}
```

Comment:
is this because they are already mounted / bound to the v1 hierarchy?
https://www.kernel.org/doc/html/latest/admin-guide/cgroup-v2.html#mounting

here is a freezer for v2, https://lwn.net/Articles/772377/
with this hardcoded assumption, can the final list of controllers end up having two "freezers"?

Reply:
"freezers" is not visible in /sys/fs/cgroup/cgroup.controllers as described above. "freezers" don' t have dfl_cftypes and the bit of cgrp_dfl_implicit_ss_mask of "freezers" is set. So the final list does not have two "freezers".

Reply:
In addition to that, "freezers" may set the flag of cgroup_subsys->implicit_on_dfl in future https://lore.kernel.org/patchwork/patch/651321/

Reply:
i think reference links should be added above these hardcoded controllers, explaining the reasoning.

Reply:
I added the reference link and explanation.